### PR TITLE
Improve webhook validation

### PR DESCRIPTION
**Pull request by dgerd**
* Add fieldmask validation for disallowed kubernetes fields
* Add validation for previously unvalidated inputs

### CODE REVIEW

File: pkg/apis/serving/v1alpha1/revision_validation.go [hunk 1]
```diff
@@ -184,120 +190,193 @@ func ValidateContainerConcurrency(cc RevisionContainerConcurrencyType, cm Revisi
...
-	if err := validateProbe(container.LivenessProbe).ViaField("livenessProbe"); err != nil {
-		errs = errs.Also(err)
-	}
-	if _, err := name.ParseReference(container.Image, name.WeakValidation); err != nil {
-		fe := &apis.FieldError{
-			Message: "Failed to parse image reference",
-			Paths:   []string{"image"},
-			Details: fmt.Sprintf("image: %q, error: %v", container.Image, err),
+func validateDisallowedFields(request, maskedRequest interface{}) *apis.FieldError {
+	if disallowed, err := kmp.CompareSetFields(request, maskedRequest); err != nil {
```
Comment (mattmoor) [inline pkg/apis/serving/v1alpha1/revision_validation.go:370]: Hmm, this probably returns Go fields, not JSON fields, which isn't exactly what we want :(

Comment (mattmoor) [reply pkg/apis/serving/v1alpha1/revision_validation.go:370]: Maybe not, it looks like you fixed that in the test diffs?

File: pkg/apis/serving/v1alpha1/revision_validation.go [hunk 2]
```diff
@@ -184,120 +190,193 @@ func ValidateContainerConcurrency(cc RevisionContainerConcurrencyType, cm Revisi
...
+	// Liveness Probes
+	errs = errs.Also(validateProbe(container.LivenessProbe).ViaField("livenessProbe"))
+	// Ports
+	errs = errs.Also(validateContainerPorts(container.Ports).ViaField("ports"))
+	// Readiness Probes
+	errs = errs.Also(validateProbe(container.ReadinessProbe).ViaField("readinessProbe"))
+	// Resources
+	errs = errs.Also(validateResources(&container.Resources).ViaField("resources"))
+	// SecurityContext
+	errs = errs.Also(validateSecurityContext(container.SecurityContext).ViaField("securityContext"))
```
Comment (mattmoor) [inline pkg/apis/serving/v1alpha1/revision_validation.go:292]: One thing that I don't like about this is that the fieldmask function and the validation function (here) have to stay in sync.

File: pkg/apis/serving/v1alpha1/revision_validation.go (hunk 1 above)
Comment (dgerd) [reply pkg/apis/serving/v1alpha1/revision_validation.go:370]: I fixed kmp to return JSON fields if they are present on the struct. :)

**Pull request closed without merging**